### ISSUE TITLE
HTTP client customizations

### ISSUE THREAD

**Issue by javierhonduco** [d-e-s-o/debuginfod#2]
Hi Daniel,

I am interesting in setting certain options to the HTTP client this crate uses, but currently this is not supported. Would be useful to customise stuff such as timeouts, etc. Perhaps a new constructor could be added where a client could be passed? I think that would offer the greater degree of flexibility, but perhaps this is something that could be problematic for maintenance. Alternatively, perhaps we could a way to configure timeouts as it seems like the most necessary addition as of now?

Let me know, happy to send a PR! 😄

**Comment by d-e-s-o** [comment]
Hey @javierhonduco , yeah, I currently treat the actual HTTP client in use as an implementation detail and would like to keep it that way. That being said, I think we can just introduce some kind of builder infrastructure to make construction more configurable (similar to what `reqwest` [already provides](https://docs.rs/reqwest/0.12.8/reqwest/blocking/struct.ClientBuilder.html)).

Feel free to open a pull request if you want to take a stab at the implementation.

**Comment by d-e-s-o** [comment]
Actually, perhaps the more appropriate solution would be to interface with the HTTP client via a trait only. Basically,
```rust
trait HttpClient {
  fn get(&self, url) -> /* whatever */
}

impl HttpClient for reqwest::Client {
  fn get(&self, url) -> ... {
    self.execute(Request::new(Method::GET, url))
  }
}
```
Or something like that. This way, we'd leave all configurability with the creator and would be more flexible when it comes to "backends" being used. Haven't fully thought that through, but it may be better down the line. It may also make adding [`async` support](https://github.com/d-e-s-o/debuginfod/issues/3) more straight forward.

**Comment by d-e-s-o** [comment]
I thought about it some more. I think for the time being the best thing would be to just introduce some Builder infrastructure and stick with the current `Client` as I mentioned. The trait thing we can do once it truly becomes necessary. I think the latter will be more flexible and all, but the problem I see is that now it becomes kind of weird to have a "default" client (after all, all impls are basically equal), which means that out of the box usage would be more problematic (now users have to choose an HTTP client crate etc). At least I don't see a good way around this issue right now.

I think we may still introduce such a trait for use internally, just to have a clearer defined interface of what functionality we expect the HTTP client to provide.

**Comment by javierhonduco** [comment]
Makes sense, thanks for sharing your thoughts! Will keep you updated. As of now being able to change the various timeouts and the user agent would probably go a long way. I don't expect any other changes needed for my use cases at least.

**Comment by javierhonduco** [comment]
Good news, [it seems that reqwest sets a 30s default timeout](https://github.com/seanmonstar/reqwest/blob/e2d4b14c695acaebe531af7ce4712e182278f38a/src/blocking/client.rs#L1297). We might want to make it configurable still, but it's not as high priority for me at least. I've used other HTTP clients in the past that had no timeout by default and was afraid it was the case here too.

**Comment by d-e-s-o** [comment]
I think this should have been addressed via https://github.com/d-e-s-o/debuginfod/pull/27 now. Let me know if something is still missing.
